lib/ui/painting/image_decoder.cc

```diff
@@ -122,31 +144,102 @@ static sk_sp<SkImage> ImageFromDecompressedData(
     return nullptr;
   }
 
-  return ResizeRasterImage(std::move(image), target_width, target_height, flow);
+  if (!target_width && !target_height) {
+    // The resized dimesions are the same as the intrinsic dimensions of the
+    // image. There is nothing to do.
+    return image->makeRasterImage();
+  }
+
+  auto resized_dimensions =
+      GetResizedDimensions(image->dimensions(), target_width, target_height);
+
+  return ResizeRasterImage(std::move(image), resized_dimensions, flow);
 }
 
-static sk_sp<SkImage> ImageFromCompressedData(
-    sk_sp<SkData> data,
-    std::optional<uint32_t> target_width,
-    std::optional<uint32_t> target_height,
-    const fml::tracing::TraceFlow& flow) {
+sk_sp<SkImage> ImageFromCompressedData(sk_sp<SkData> data,
+                                       std::optional<uint32_t> target_width,
+                                       std::optional<uint32_t> target_height,
+                                       const fml::tracing::TraceFlow& flow) {
   TRACE_EVENT0("flutter", __FUNCTION__);
   flow.Step(__FUNCTION__);
 
-  auto decoded_image = SkImage::MakeFromEncoded(data);
+  if (!target_width && !target_height) {
+    // The resized dimesions are the same as the intrinsic dimensions of the
+    // image. There is nothing to do.
+    return SkImage::MakeFromEncoded(data)->makeRasterImage();
+  }
 
-  if (!decoded_image) {
+  auto codec = SkCodec::MakeFromData(data);
+  FML_DCHECK(codec);
+  if (codec == nullptr) {
+    FML_LOG(ERROR) << "Failed to instantiate image decoder.";
     return nullptr;
   }
 
-  // Make sure to resolve all lazy images.
-  decoded_image = decoded_image->makeRasterImage();
+  const auto* codec_ptr = codec.get();
```

Comment:
Not sure why this is necessary. sk_sp already has the operator overloaded.

Reply:
I assume you mean the dereference/pointer operators. Note that codec is a unique_ptr (not shared) and I have to move it into the ImageGenerator however, I need to call the getScaledDimensions later once I have figured out what the source width and height are respecting the orientation. Otherwise, I'd just get the dimensions from the codec.